File: scripts/merge_substitution_templates.py
#!/usr/bin/env python3
"""Merge substitution templates into equation sheet JSON files."""
import json
from pathlib import Path

root = Path(__file__).resolve().parent.parent
sheets_dir = root / "data" / "equation_sheets"
source = json.loads((sheets_dir / "substitution_templates.json").read_text(encoding="utf-8"))
templates = source.get("templates", {})
skip_ids = set(source.get("_skip_ids", []))

merged_count = 0
for file_path in sorted(sheets_dir.glob("*.json")):
    if file_path.name == "substitution_templates.json":
        continue
    data = json.loads(file_path.read_text(encoding="utf-8-sig"))
    if not isinstance(data.get("equations"), list):
        continue
    changed = False
    new_equations = []
    for eq in data["equations"]:
        tpl = templates.get(eq.get("id"))
        if not tpl or eq.get("id") in skip_ids:
            new_equations.append(eq)
            continue
        changed = True
        merged_count += 1
        merged = {**eq}
        if "substitution_template" in tpl:
            merged["substitution_template"] = tpl["substitution_template"]
        if "rearrangement_forms" in tpl:
            merged["rearrangement_forms"] = tpl["rearrangement_forms"]
        new_equations.append(merged)
    if changed:
        data["equations"] = new_equations
        file_path.write_text(json.dumps(data, indent=2) + "\n", encoding="utf-8")
        print(f"Updated {file_path.name}")

print(f"Merged templates into {merged_count} equation entries.")

File: scripts/merge_substitution_templates.mjs
#!/usr/bin/env node
/**
 * Merge substitution_template + rearrangement_forms from
 * data/equation_sheets/substitution_templates.json into all equation sheet JSON files.
 */
import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const root = path.join(__dirname, "..");
const sheetsDir = path.join(root, "data", "equation_sheets");
const templatesPath = path.join(sheetsDir, "substitution_templates.json");

const source = JSON.parse(fs.readFileSync(templatesPath, "utf8"));
const templates = source.templates || {};
const skipIds = new Set(source._skip_ids || []);

const files = fs.readdirSync(sheetsDir).filter(
  (f) => f.endsWith(".json") && f !== "substitution_templates.json"
);

let mergedCount = 0;
for (const file of files) {
  const filePath = path.join(sheetsDir, file);
  const data = JSON.parse(fs.readFileSync(filePath, "utf8"));
  if (!Array.isArray(data.equations)) continue;

  let changed = false;
  data.equations = data.equations.map((eq) => {
    const tpl = templates[eq.id];
    if (!tpl || skipIds.has(eq.id)) return eq;
    changed = true;
    mergedCount++;
    return {
      ...eq,
      substitution_template: tpl.substitution_template,
      rearrangement_forms: tpl.rearrangement_forms
    };
  });

  if (changed) {
    fs.writeFileSync(filePath, JSON.stringify(data, null, 2) + "\n");
    console.log(`Updated ${file}`);
  }
}

console.log(`Merged templates into ${mergedCount} equation entries across ${files.length} files.`);
